src/xrpld/ledger/detail/View.cpp

```diff
@@ -1923,24 +2175,141 @@ requireAuth(
 {
     auto const mptID = keylet::mptIssuance(mptIssue.getMptID());
     auto const sleIssuance = view.read(mptID);
-
     if (!sleIssuance)
         return tecOBJECT_NOT_FOUND;
 
     auto const mptIssuer = sleIssuance->getAccountID(sfIssuer);
 
     // issuer is always "authorized"
-    if (mptIssuer == account)
+    if (mptIssuer == account)  // Issuer won't have MPToken
         return tesSUCCESS;
 
     auto const mptokenID = keylet::mptoken(mptID.key, account);
     auto const sleToken = view.read(mptokenID);
+    // Note, this is not amendment-gated because  we do not want to maintain in
+    // this file the list of all the amendments which can write to this field.
+    // Without additional amendments this field is always empty.
+    auto const maybeDomainID = sleIssuance->at(~sfDomainID);
+    if (!maybeDomainID)
+    {
+        // if account has no MPToken, fail
+        if (!sleToken)
+            return tecNO_AUTH;
+
+        // mptoken must be authorized if issuance enabled requireAuth
+        if (sleIssuance->getFieldU32(sfFlags) & lsfMPTRequireAuth &&
+            !(sleToken->getFlags() & lsfMPTAuthorized))
+            return tecNO_AUTH;
+
+        return tesSUCCESS;
+    }
+
+    // err = tefINTERNAL | tecINVALID_DOMAIN | tecNO_PERMISSION | tecEXPIRED
+    if (auto const err =
+            credentials::validDomain(view, *maybeDomainID, account);
+        !isTesSuccess(err))
+        return err;
+
+    // We are authorized by permissioned domain.
+    return tesSUCCESS;
+}
+
+[[nodiscard]] TER
+enforceMPTokenAuthorization(
```

Comment:
Is this function to be used anywhere that one needs to check if a MPToken is authorized? Or is this only used by Vault?

I do a similar check in escrow, I will do a similar one in paychan.